### PR TITLE
Bulk Mapping Function example

### PR DESCRIPTION
**Pull request by Shwetabhk**

### CODE REVIEW

File: apps/mappings/tasks.py [hunk 1]
```diff
@@ -17,6 +17,43 @@
...
+    :param project_attributes: Destination Attributes List
+    :param workspace_id: workspace_id
+    :return: mappings list
+    """
+    project_names = []
+
+    for project in project_attributes:
+        project_names.append(project.value)
+
+    source_attributes: List[ExpenseAttribute] = ExpenseAttribute.objects.filter(
```
Comment (Shwetabhk) [inline apps/mappings/tasks.py:32]: This query will get all the source attributes which are unmapped and in the list of project_names

Comment (Shwetabhk) [reply apps/mappings/tasks.py:32]: Query 1

File: apps/mappings/tasks.py [hunk 2]
```diff
@@ -17,6 +17,43 @@
...
+        value__in=project_names, workspace_id=workspace_id, mapping__source_id_isnull=True).all()
+
+    source_value_id_map = {}
+
+    for source_attribute in source_attributes:
```
Comment (Shwetabhk) [inline apps/mappings/tasks.py:37]: creating a map of all the value and id

File: apps/mappings/tasks.py [hunk 3]
```diff
@@ -17,6 +17,48 @@
...
+    Bulk create mappings
+    :param destination_type: Destination Type
+    :param source_type: Source Type
+    :param destination_attributes: Destination Attributes List
+    :param workspace_id: workspace_id
+    :return: mappings list
+    """
+    attribute_value_list = []
+
+    for destination_attribute in destination_attributes:
```
Comment (ashwin1111) [inline apps/mappings/tasks.py:32]: `attribute_value_list = destination_attributes.values_list('value', flat=True)`

File: apps/mappings/tasks.py [hunk 4]
```diff
@@ -17,6 +17,48 @@
...
+        value__in=attribute_value_list, workspace_id=workspace_id, mapping__source_id__isnull=True).all()
+
+    source_value_id_map = {}
+
+    for source_attribute in source_attributes:
+        source_value_id_map[source_attribute.value.lower()] = source_attribute.id
+
+    mapping_batch = []
+
+    for detination_attribute in destination_attributes:
```
Comment (ashwin1111) [inline apps/mappings/tasks.py:45]: `destination_attribute`

File: apps/mappings/tasks.py (hunk 3 above)
Comment (Shwetabhk) [reply apps/mappings/tasks.py:36]: This we can do when destination_attributes will be a queryset, right now it is not

Comment (Shwetabhk) [reply apps/mappings/tasks.py:36]: right now it is just a list

**Review by ashwin1111 — APPROVED** [review]

File: apps/mappings/tasks.py [hunk 5]
```diff
@@ -17,6 +17,47 @@
 logger = logging.getLogger(__name__)
 
 
+def bulk_create_mappings(destination_attributes: List[DestinationAttribute], source_type: str,
```
Comment (ashwin1111) [inline apps/mappings/tasks.py:20]: `destination_attributes: list`

File: apps/mappings/tasks.py [hunk 6]
```diff
@@ -235,36 +242,9 @@ def auto_create_category_mappings(workspace_id):
...
-                    'source_value': category.value,
-                    'destination_value': category.value
-                }
-                logger.error(
-                    'Error while creating categories auto mapping workspace_id - %s %s',
-                    workspace_id, {'payload': detail}
-                )
-                raise ExpenseAttribute.DoesNotExist
-
+        category_mappings = bulk_create_mappings(fyle_categories, 'CATEOGORY', 'ACCOUNT', workspace_id)
```
Comment (Sravanksk) [inline apps/mappings/tasks.py:247]: typo - `CATEGORY`

**Review by Sravanksk — APPROVED** [review]

**Review by ashwin1111 — APPROVED** [review]

**Pull request merged**